### PR TITLE
Collect trace data through the observatory HTTP interface

### PR DESCRIPTION
**Pull request by jason-simmons**

### CODE REVIEW

**Comment by eseidelGoogle** [comment]
@johnmccutchan or @devoncarew are probably your best reviewers.  I checked and both web_socket_channel and json_rpc_2 were developed by the Dart team/Google, so there is no concern re: licenses of those dependencies.

File: packages/flutter_tools/lib/src/android/android_device.dart [hunk 1]
```diff
@@ -370,59 +362,58 @@ class AndroidDevice extends Device {
...
-        'Did you remember to start the trace before stopping it?');
-    return null;
+
+    await client.sendRequest('_setVMTimelineFlags', {'recordedStreams': '[]'});
+
+    String localPath;
+    if (outPath != null) {
+      localPath = outPath;
+    } else {
+      String timestamp = new DateFormat('yyyy_MM_dd_hh_mm_ss').format(new DateTime.now());
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/android/android_device.dart]: If it's not critical that the file includes the current date and time, we could write the file out as `trace_xx.json` here. We use that format for crash reports and for screenshots. There's a utility method (https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/base/utils.dart#L44) to create a unique file w/ that name pattern.

File: packages/flutter_tools/lib/src/android/android_device.dart [hunk 2]
```diff
@@ -370,59 +362,58 @@ class AndroidDevice extends Device {
...
+      localPath = 'trace_$timestamp.json';
+    }
+
+    Map<String, dynamic> response = await client.sendRequest('_getVMTimeline');
+    List<dynamic> traceEvents = response['traceEvents'];
+
+    IOSink sink = new File(localPath).openWrite();
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/android/android_device.dart]: I _think_ the below code can be replaced by `new File(localPath).writeAsBytesSync(JSON.encode(traceEvents));`

File: packages/flutter_tools/pubspec.yaml [hunk 3]
```diff
@@ -13,14 +13,17 @@ dependencies:
...
+  intl: '>=0.12.4+2 <0.13.0'
+  json_rpc_2: ^2.0.0
   json_schema: ^1.0.3
   mustache4dart: ^1.0.0
   package_config: ^0.1.3
   path: ^1.3.0
   pub_semver: ^1.0.0
   stack_trace: ^1.4.0
-  yaml: ^2.1.3
+  web_socket_channel: ^1.0.0
```
Comment (devoncarew) [inline packages/flutter_tools/pubspec.yaml]: I'm hoping to tie into the observatory from `flutter run` shortly, at which point I'll probably use the https://github.com/dart-lang/vm_service_drivers/tree/master/dart library. @yjbanov is using the https://github.com/dart-lang/vm_service_client library in flutter_driver. I don't know that this is actionable in this PR, but at some point we should look at consolidating.

**Comment by devoncarew** [comment]
lgtm, though I think we should simplify the file name generating; it'll be similar to other places where we create files.

File: packages/flutter_tools/lib/src/android/android_device.dart (hunk 1 above)
Comment (jason-simmons) [inline packages/flutter_tools/lib/src/android/android_device.dart]: done

File: packages/flutter_tools/lib/src/android/android_device.dart (hunk 2 above)
Comment (jason-simmons) [inline packages/flutter_tools/lib/src/android/android_device.dart]: The goal of that code is to stream the JSON to a file without buffering the entire JSON string in memory

Comment (devoncarew) [inline packages/flutter_tools/lib/src/android/android_device.dart]: Ah, ok.

**Comment by devoncarew** [comment]
lgtm!

@jason-simmons, my reading of this code now is that it's no longer android specific, and could work with iOS as well. Does that sounds right?

File: packages/flutter_tools/lib/src/android/android_device.dart [hunk 4]
```diff
@@ -370,59 +363,57 @@ class AndroidDevice extends Device {
...
-        throw 'Error code $exitCode returned when running ${catCommand.join(" ")}';
-
-      runSync(adbCommandForDevice(
-          <String>['shell', 'run-as', apk.id, 'rm', tracePath]
-      ));
-      return localPath;
+    await client.sendRequest('_setVMTimelineFlags',
+        {'recordedStreams': ['Compiler', 'Dart', 'Embedder', 'GC']}
+    );
+    await client.sendRequest('_clearVMTimeline');
```
Comment (johnmccutchan) [inline packages/flutter_tools/lib/src/android/android_device.dart]: Do we always want to clear?

**Comment by johnmccutchan** [comment]
lgtm

**Comment by chinmaygarde** [comment]
> my reading of this code now is that it's no longer android specific, and could work with iOS as well.

Yes! Thanks Jason.

**Pull request merged**